urllib3/response.py

```diff
@@ -56,14 +56,30 @@ class GzipDecoder(object):
 
     def __init__(self):
         self._obj = zlib.decompressobj(16 + zlib.MAX_WBITS)
+        self._state = 'first'
 
     def __getattr__(self, name):
         return getattr(self._obj, name)
 
     def decompress(self, data):
+        ret = binary_type()
+        if self._state == 'error':
+            return ret
         if not data:
             return data
-        return self._obj.decompress(data)
+        try:
+            ret = self._obj.decompress(data)
+            while self._obj.unused_data:
+                self._state = 'other'
+                (self._obj, data) = (zlib.decompressobj(16 + zlib.MAX_WBITS),
+                                     self._obj.unused_data)
+                ret += self._obj.decompress(data)
+            return ret
+        except Exception as ex:
```

Comment:
Let's catch the specific class of exception we expect and not swallow everything. Always remember to keep try/except blocks as narrow as possible both in terms of lines-of-code covered and class-of-error covered. `try/except/else` is a useful construct to use.

Reply:
Done.  But the exception catching is still essentially on the whole function.  They are for the two .decompress calls, either can cause the exception, and the handling is the same.  The other code has no potential to cause exceptions, so there is no harm doing it the current way.